### PR TITLE
ARROW-12388: [C++][Gandiva] Implement cast numbers from varbinary functions in gandiva

### PR DESCRIPTION
**Pull request by jvictorhuguenin**

### CODE REVIEW

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 1]
```diff
@@ -307,6 +306,37 @@ CAST_NUMERIC_FROM_STRING(double, arrow::DoubleType, FLOAT8)
 
 #undef CAST_NUMERIC_FROM_STRING
 
+#define CAST_NUMERIC_FROM_VARBINARY(OUT_TYPE, ARROW_TYPE, TYPE_NAME)                   \
+  GANDIVA_EXPORT                                                                       \
+  OUT_TYPE gdv_fn_cast##TYPE_NAME##_varbinary(int64_t context, const char* data,       \
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:311]: Is it any different from CAST_NUMERIC_FROM_STRING functions. Can't you just call them

Comment (projjal) [reply cpp/src/gandiva/gdv_function_stubs.cc:311]: Use a macro to use the same definition for both string and varbinary inputs

Comment (frank400) [reply cpp/src/gandiva/gdv_function_stubs.cc]: done

**Review by projjal — APPROVED** [review]

**Comment by projjal** [comment]
Can you look at the failing tests

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 2]
```diff
@@ -300,12 +299,17 @@ char* gdv_fn_dec_to_string(int64_t context, int64_t x_high, uint64_t x_low,
     return val;                                                                      \
   }
 
-CAST_NUMERIC_FROM_STRING(int32_t, arrow::Int32Type, INT)
-CAST_NUMERIC_FROM_STRING(int64_t, arrow::Int64Type, BIGINT)
-CAST_NUMERIC_FROM_STRING(float, arrow::FloatType, FLOAT4)
-CAST_NUMERIC_FROM_STRING(double, arrow::DoubleType, FLOAT8)
+CAST_NUMERIC_STRING(int32_t, arrow::Int32Type, INT, utf8)
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:302]: Can you change the macros to CAST_NUMERIC_FROM_VARLEN_TYPES, another two macros CAST_NUMERIC_FROM_STRING and CAST_NUMERIC_FROM_BINARY which calls the above macro with utf8 or binary. Call these two macros with numberic types

File: cpp/src/gandiva/gdv_function_stubs_test.cc [hunk 3]
```diff
@@ -290,4 +290,140 @@ TEST(TestGdvFnStubs, TestCastVARCHARFromDouble) {
...
+}
+
+TEST(TestGdvFnStubs, TestCastVarbinaryBIGINT) {
+  gandiva::ExecutionContext ctx;
+
+  int64_t ctx_ptr = reinterpret_cast<int64_t>(&ctx);
+
+  EXPECT_EQ(gdv_fn_castBIGINT_varbinary(ctx_ptr, "-45", 3), -45);
+  EXPECT_EQ(gdv_fn_castBIGINT_varbinary(ctx_ptr, "0", 1), 0);
+  EXPECT_EQ(gdv_fn_castBIGINT_varbinary(ctx_ptr, "9223372036854775807", 19),
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:344]: Can you add some tests with hex string containing valid numbers

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 2 above)
Comment (jvictorhuguenin) [reply cpp/src/gandiva/gdv_function_stubs.cc]: done

File: cpp/src/gandiva/gdv_function_stubs_test.cc (hunk 3 above)
Comment (jvictorhuguenin) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:345]: done

**Comment by praveenbingo** [comment]
@projjal can you please confirm the errors on the builds is not related.

**Comment by projjal** [comment]
> @projjal can you please confirm the errors on the builds is not related.

Build failures are unrelated

**Comment by praveenbingo** [comment]
@projjal Needs a rebase

**Comment by projjal** [comment]
@jvictorhuguenin Can you rebase. There is merge conflict.

**Comment by kou** [comment]
@praveenbingo Could you use the latest `dev/merge_arrow_pr.py`? The latest `dev/merge_arrow_pr.py` doesn't list squashed commits.

**Comment by praveenbingo** [comment]
@kou Will pull the latest in. Thanks - sorry if my merge caused any trouble.

**Comment by kou** [comment]
Thanks!

**Pull request closed without merging**